Makefile

```diff
@@ -1,205 +1,468 @@
+	$(call bold_msg, "EXTRACT", "STAGE 2 PLUGINS")
+	env \
+	  SRC=ulib/ \
+	  FSTAR_EXE=$(FSTAR2_BARE_EXE) \
+	  CACHE_DIR=stage2/plugins.checked/ \
+	  OUTPUT_DIR=stage2/plugins.ml/ \
+	  CODEGEN=PluginNoLib \
+	  OTHERFLAGS="--ext __guts $(OTHERFLAGS)" \
+	  TAG=plugins \
+	  $(MAKE) -f mk/plugins.mk ocaml
+
+$(FSTAR2_FULL_EXE): $(FSTAR2_FULL_EXE).src _force
+	$(call bold_msg, "BUILD", "STAGE 2 FSTARC")
+	$(MAKE) -C stage2 fstarc-full FSTAR_DUNE_RELEASE=1
+
+2.alib.src: $(FSTAR2_FULL_EXE) _force
+	$(call bold_msg, "EXTRACT", "STAGE 2 LIB")
+	env \
+	  SRC=ulib/ \
+	  FSTAR_EXE=$(FSTAR2_FULL_EXE) \
+	  CACHE_DIR=stage2/ulib.checked/ \
+	  OUTPUT_DIR=stage2/ulib.ml/ \
+	  CODEGEN=OCaml \
+	  TAG=lib \
+	  $(MAKE) -f mk/lib.mk all
+
+2.alib: 2.alib.src _force
+	$(call bold_msg, "BUILD", "STAGE 2 LIB")
+	$(MAKE) -C stage2/ libapp FSTAR_DUNE_RELEASE=1
+
+2.plib.src: $(FSTAR2_FULL_EXE) 2.alib.src _force
+	# NB: shares .depend and checked from 2.alib.src,
+	# hence the dependency, though it is not quite precise.
+	$(call bold_msg, "EXTRACT", "STAGE 2 PLUGLIB")
+	env \
+	  SRC=ulib/ \
+	  FSTAR_EXE=$(FSTAR2_FULL_EXE) \
+	  CACHE_DIR=stage2/ulib.checked/ \
+	  OUTPUT_DIR=stage2/ulib.pluginml/ \
+	  CODEGEN=PluginNoLib \
+	  TAG=pluginlib \
+	  DEPFLAGS='--extract +FStar.Tactics,+FStar.Reflection,+FStar.Sealed' \
+	  $(MAKE) -f mk/lib.mk all
+
+2.plib: 2.plib.src _force | 2.alib # this last dependency only to prevent simultaneous dune builds
+	$(call bold_msg, "BUILD", "STAGE 2 PLUGLIB")
+	$(MAKE) -C stage2/ libplugin FSTAR_DUNE_RELEASE=1
+
+# Stage 3 is different, we don't build it, we just check that the
+# extracted OCaml files coincide exactly with stage2. We also do not
+# extract the plugins, as is stage2/fstarc and stage3/fstarc coincide,
+# then they are exactly the same compiler and will extract the plugins
+# in the same way.
+
+stage3-bare: $(FSTAR2_FULL_EXE) _force
+	$(call bold_msg, "EXTRACT", "STAGE 3 FSTARC")
+	# NOTE: see the explanation for FSTAR_LIB near top of file.
+	env \
+	  SRC=src/ \
+	  FSTAR_EXE=$(FSTAR2_FULL_EXE) \
+	  FSTAR_LIB=$(abspath ulib) \
+	  CACHE_DIR=stage3/fstarc.checked/ \
+	  OUTPUT_DIR=stage3/fstarc.ml/ \
+	  CODEGEN=OCaml \
+	  TAG=fstarc \
+	  $(MAKE) -f mk/fstar-12.mk ocaml
+
+stage3-diff: stage3-bare _force
+	$(call bold_msg, "DIFF", "STAGE 2 vs STAGE 3")
+	@# No output expected the gitignore line
+	diff -r stage2/fstarc.ml stage3/fstarc.ml
+
+$(INSTALLED_FSTAR1_FULL_EXE): 1.full 1.alib.src 1.plib.src
+	$(call bold_msg, "INSTALL", "STAGE 1")
+	$(MAKE) -C stage1 install
+
+$(INSTALLED_FSTAR2_FULL_EXE): 2.full 2.alib.src 2.plib.src
+	$(call bold_msg, "INSTALL", "STAGE 2")
+	$(MAKE) -C stage2 install FSTAR_DUNE_RELEASE=1
+
+setlink-%:
+	if [ -e out ] && ! [ -h out ]; then echo "ERROR: out/ exists and is not a symbolic link, please remove it"; false; fi
+	ln -Trsf stage$*/out out
+	# For compatibility with the previous layout
+	mkdir -p bin
+	ln -Trsf out/bin/fstar.exe bin/fstar.exe
```

Comment:
I still think it's a bad idea to make the default F* binary point to whatever make target succeeded last, and this will likely confuse the hell out of me.

It's great that you changed the ulib vscode config to point to stage1 directly (and src to stage0).  That addresses most of my practical concerns with debugging failing builds.

Reply:
I tend to agree, but I think we'll want a way to be able to run the tests on the stage 1 compiler. Currently that's as easy as `make 1 && make test`. Note that the `out` link is *only* created when you call `make 1` or `make 2`. Most users will not use these rules and just get the stage 2 compiler, i.e. the actual final product.

I'd also be happy having a differently named rule for setting the link to the stage 1, `fake-1` or whatever.

Reply:
I think the most natural way to run tests is `make test`; and I'd expect that it would be roughly similar to CI, i.e., it should build F* and run the tests on stage 2.

I also just realized that `make test` by itself would fail (or use an out-of-date fstar.exe) since it doesn't depend on any build target.  That's really unexpected.

Another options is to have `test-1` and `test-2` targets, and maybe `test: test-2` for convenience.  (Then the tests could also use different cache directories for each stage if we want to support that.)

> Most users will not use these rules

What you're saying is that if you only ever call `make`, then this defaults to `make 2` and you'll only ever get a symlink to stage 2, right?

Is there a target to build stage N without messing with the symlinks?

Reply:
> I think the most natural way to run tests is `make test`; and I'd expect that it would be roughly similar to CI, i.e., it should build F* and run the tests on stage 2.
> 
> I also just realized that `make test` by itself would fail (or use an out-of-date fstar.exe) since it doesn't depend on any build target. That's really unexpected.
> 
> Another options is to have `test-1` and `test-2` targets, and maybe `test: test-2` for convenience. (Then the tests could also use different cache directories for each stage if we want to support that.)

I changed the rules to roughly do this (see last patch). I kept a `make _test` available that will just use the latest one, since I think I will be running this often when working with a stage 1 (to avoid the overhead of dune realizing there's nothing to do).

> 
> > Most users will not use these rules
> 
> What you're saying is that if you only ever call `make`, then this defaults to `make 2` and you'll only ever get a symlink to stage 2, right?

Yes exactly.

> 
> Is there a target to build stage N without messing with the symlinks?

I added `stage1` and `stage2` to build the stages without setting the `out/` link. The shortcuts `1` and `2` call the respective stage target and then set the link.

Reply:
SGTM.  Thank you!